distsql: support lookup join on secondary index

solongordon opened the pull request:
joinReader now supports lookup joins on secondary indexes. This was a
trivial change for queries where all the output columns are included in
the secondary index. I just modified the physical planner to specify the
secondary index in the JoinReaderSpec and removed checks which prevented
secondary indexes from being used.

The more complicated situation is when we want to do a lookup join
against a non-covering index. In this case, the logical planner plans an
index join before the inner join, but we want to perform the lookup join
first. We now handle this by only planning the lookup join during
physical planning, not the index join. During execution, the joinReader
detects that there are output columns not covered by the secondary
index, and it performs primary index lookups as necessary to retrieve
the additional columns.

Fixes #25431

solongordon committed:
distsql: support lookup join on secondary index

joinReader now supports lookup joins on secondary indexes. This was a
trivial change for queries where all the output columns are included in
the secondary index. I just modified the physical planner to specify the
secondary index in the JoinReaderSpec and removed checks which prevented
secondary indexes from being used.

The more complicated situation is when we want to do a lookup join
against a non-covering index. In this case, the logical planner plans an
index join before the inner join, but we want to perform the lookup join
first. We now handle this by only planning the lookup join during
physical planning, not the index join. During execution, the joinReader
detects that there are output columns not covered by the secondary
index, and it performs primary index lookups as necessary to retrieve
the additional columns.

Fixes #25431

Release note (sql change): The experimental lookup join feature now
supports secondary indexes.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/25628)
<!-- Reviewable:end -->

solongordon commented:
cc @asubiotto @petermattis 

I'm also planning to add tests to `joinreader_test.go` but wanted to get some eyes on this.

solongordon commented:
Is it acceptable not to add a new version for this feature since lookup joins are still experimental?

---

Review status: 0 of 3 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/25628#-:-LCjfZsW6u4R3owS-k_J:b-ee9e2b)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Nice, this looks easier than I though it would be!

I would add a version, but I wouldn't bump the min-accepted version.

---

Review status: 0 of 3 files reviewed at latest revision, all discussions resolved.

---

*[pkg/sql/distsqlrun/joinreader.go, line 168 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/25628#-LCjfcCyCGDM4iTRQVux:-LCjfcCyCGDM4iTRQVuy:bif73uj) ([raw file](https://github.com/cockroachdb/cockroach/blob/79776afe24b0ed005a2a04b5bebb4f3b1e813e23/pkg/sql/distsqlrun/joinreader.go#L168)):*
> ```Go
> ) bool {
> 	coveredCols := util.MakeFastIntSet()
> 	for _, i := range append(index.ColumnIDs, index.ExtraColumnIDs...) {
> ```

I don't understand what we're doing here. Aren't `neededColumns` ordinals (0, 1, 2, 3..) and not column IDs?

Also, don't mutate the slice that belongs to `index` just to save a line of code. What if it has spare capacity and we have two separate queries running this code in parallel?

---

*[pkg/sql/distsqlrun/joinreader.go, line 171 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/25628#-LCjfMwpD7SZHwdSWINw:-LCjfMwpD7SZHwdSWINx:bvmx0p8) ([raw file](https://github.com/cockroachdb/cockroach/blob/79776afe24b0ed005a2a04b5bebb4f3b1e813e23/pkg/sql/distsqlrun/joinreader.go#L171)):*
> ```Go
> 		coveredCols.Add(numInputCols + int(i) - 1)
> 	}
> 	for _, col := range neededColumns.Ordered() {
> ```

return `neededColumns.SubsetOf(coveredCols)`

---

*[pkg/sql/distsqlrun/joinreader.go, line 179 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/25628#-LCjgS0H-b3eZOKrSHEc:-LCjgS0H-b3eZOKrSHEd:b-8thz74) ([raw file](https://github.com/cockroachdb/cockroach/blob/79776afe24b0ed005a2a04b5bebb4f3b1e813e23/pkg/sql/distsqlrun/joinreader.go#L179)):*
> ```Go
> }
> 
> func getColumnIndexes(index sqlbase.IndexDescriptor) util.FastIntSet {
> ```

[nit] indices (we use "indexes" for table indexes)

---

*[pkg/sql/distsqlrun/joinreader.go, line 182 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/25628#-LCjgEZN1KL5JGmE8uzX:-LCjgEZN1KL5JGmE8uzY:bpp043e) ([raw file](https://github.com/cockroachdb/cockroach/blob/79776afe24b0ed005a2a04b5bebb4f3b1e813e23/pkg/sql/distsqlrun/joinreader.go#L182)):*
> ```Go
> 	cols := util.MakeFastIntSet()
> 	for _, col := range index.ColumnIDs {
> 		// Convert from 1-based to 0-based.
> ```

Same problem here, you are mapping Column IDs to ordinals. This perhaps happens to work because the table you are testing with didn't have schema changes, but won't work in general.

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/25628#-:-LCjgd2U7S-iASAy5ZaE:biccfwp)*
<!-- Sent from Reviewable.io -->

solongordon commented:
Review status: 0 of 3 files reviewed at latest revision, 4 unresolved discussions.

---

*[pkg/sql/distsqlrun/joinreader.go, line 182 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/25628#-LCjgEZN1KL5JGmE8uzX:-LCjji1_Bu-ljV4qfV6Y:b73hdos) ([raw file](https://github.com/cockroachdb/cockroach/blob/79776afe24b0ed005a2a04b5bebb4f3b1e813e23/pkg/sql/distsqlrun/joinreader.go#L182)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Same problem here, you are mapping Column IDs to ordinals. This perhaps happens to work because the table you are testing with didn't have schema changes, but won't work in general.
</blockquote></details>

Got it, thank you for clarifying that. What is the preferred way to map Column ID to ordinal? Iterate over `TableDescriptor.Columns`?

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/25628)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Review status: 0 of 3 files reviewed at latest revision, 4 unresolved discussions.

---

*[pkg/sql/distsqlrun/joinreader.go, line 182 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/25628#-LCjgEZN1KL5JGmE8uzX:-LCjmGkq6AcDnoHRftmw:bbsueq) ([raw file](https://github.com/cockroachdb/cockroach/blob/79776afe24b0ed005a2a04b5bebb4f3b1e813e23/pkg/sql/distsqlrun/joinreader.go#L182)):*
<details><summary><i>Previously, solongordon (Solon) wrote…</i></summary><blockquote>

Got it, thank you for clarifying that. What is the preferred way to map Column ID to ordinal? Iterate over `TableDescriptor.Columns`?
</blockquote></details>

`initRowFetcher` builds a `colIdxMap`, perhaps that should be pulled out (especially since we are now calling it twice in some cases).

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/25628)*
<!-- Sent from Reviewable.io -->

solongordon commented:
Review status: 0 of 9 files reviewed at latest revision, 7 unresolved discussions.

---

*[pkg/sql/distsql_physical_planner.go, line 802 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/25628#-LCjyGPF6xh1ohlaOGO9:-LCk1iWp037-kZypT8ux:b-3lj3hs) ([raw file](https://github.com/cockroachdb/cockroach/blob/79776afe24b0ed005a2a04b5bebb4f3b1e813e23/pkg/sql/distsql_physical_planner.go#L802)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

I think you probably should be using the index ID (i.e. `n.index.ID`). I don't think there are any guarantees that the positional index of a SQL index within `n.desc.Indexes` is immutable. There is a guarantee that the index ID will not change. 
</blockquote></details>

Hm, this was just a refactor so I could pull out logic for getting an `IndexIdx`. If this behavior is dangerous than probably we should open a separate issue?

---

*[pkg/sql/distsqlrun/joinreader.go, line 65 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/25628#-LCjyvPxC2La8SrLdlAC:-LCk3Jx_EUaVPQ0t079c:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/79776afe24b0ed005a2a04b5bebb4f3b1e813e23/pkg/sql/distsqlrun/joinreader.go#L65)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

`sizeof(sqlbase.RowFetcher) == 328`. I think this should be a pointer that you only initialize of a primary lookup is required. This would also allow getting rid of the `primaryLookupRequired` field.
</blockquote></details>

Done.

---

*[pkg/sql/distsqlrun/joinreader.go, line 168 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/25628#-LCjfcCyCGDM4iTRQVux:-LCjw0FhB7IroyNVtUOo:brmypa2) ([raw file](https://github.com/cockroachdb/cockroach/blob/79776afe24b0ed005a2a04b5bebb4f3b1e813e23/pkg/sql/distsqlrun/joinreader.go#L168)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

I don't understand what we're doing here. Aren't `neededColumns` ordinals (0, 1, 2, 3..) and not column IDs?

Also, don't mutate the slice that belongs to `index` just to save a line of code. What if it has spare capacity and we have two separate queries running this code in parallel?
</blockquote></details>

Done. I was fuzzy on the ordinal vs ID distinction but now I've seen the light.

---

*[pkg/sql/distsqlrun/joinreader.go, line 171 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/25628#-LCjfMwpD7SZHwdSWINw:-LCjw6VhDIqjI4V1kV7b:b4xxbr5) ([raw file](https://github.com/cockroachdb/cockroach/blob/79776afe24b0ed005a2a04b5bebb4f3b1e813e23/pkg/sql/distsqlrun/joinreader.go#L171)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

return `neededColumns.SubsetOf(coveredCols)`
</blockquote></details>

Done. I had to add in the input cols to `coveredCols` for this to work.

---

*[pkg/sql/distsqlrun/joinreader.go, line 179 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/25628#-LCjgS0H-b3eZOKrSHEc:-LCjwQd12_xI3afEw8mw:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/79776afe24b0ed005a2a04b5bebb4f3b1e813e23/pkg/sql/distsqlrun/joinreader.go#L179)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

[nit] indices (we use "indexes" for table indexes)
</blockquote></details>

Done.

---

*[pkg/sql/distsqlrun/joinreader.go, line 182 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/25628#-LCjgEZN1KL5JGmE8uzX:-LCjwRhW1oB-CFkUP9_k:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/79776afe24b0ed005a2a04b5bebb4f3b1e813e23/pkg/sql/distsqlrun/joinreader.go#L182)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

`initRowFetcher` builds a `colIdxMap`, perhaps that should be pulled out (especially since we are now calling it twice in some cases).
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/25628)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
*[pkg/sql/distsql_physical_planner.go, line 802 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/25628#-LCjyGPF6xh1ohlaOGO9:-LCk3pmL2rmQCn1r9TuJ:b-q52vpl) ([raw file](https://github.com/cockroachdb/cockroach/blob/79776afe24b0ed005a2a04b5bebb4f3b1e813e23/pkg/sql/distsql_physical_planner.go#L802)):*
<details><summary><i>Previously, solongordon (Solon) wrote…</i></summary><blockquote>

Hm, this was just a refactor so I could pull out logic for getting an `IndexIdx`. If this behavior is dangerous than probably we should open a separate issue?
</blockquote></details>

We are putting whatever version of `n.desc` we have in the spec (above); `distsqlrun` does not look up table descriptors.

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/25628)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
LGTM modulo some comments.

Still need to add a version.

---

Reviewed 1 of 7 files at r2.
Review status: 1 of 9 files reviewed at latest revision, 3 unresolved discussions.

---

*[pkg/sql/distsql_physical_planner.go, line 1831 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/25628#-LCjyTwY1lK4cZ41eV6m:-LCmrEx18Axm8rTpWvNK:bx4teut) ([raw file](https://github.com/cockroachdb/cockroach/blob/79776afe24b0ed005a2a04b5bebb4f3b1e813e23/pkg/sql/distsql_physical_planner.go#L1831)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Do we need to check for the inverse scenario, where the right side of the join can be used to "lookup" in the left side? Or is that already checked for via some other means? @raduberinde?
</blockquote></details>

I don't have all the context here, but perhaps the idea was that the query can be rewritten to flip the order if necessary. If we want to handle the symmetric case, it should be a separate PR.

---

*[pkg/sql/distsqlrun/joinreader.go, line 64 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/25628#-LCmtUX27OL1CxSDBIqt:-LCmtUX27OL1CxSDBIqu:b-ayjlvs) ([raw file](https://github.com/cockroachdb/cockroach/blob/27608f8ae412589cb9d6510cae3245fe62b9dad7/pkg/sql/distsqlrun/joinreader.go#L64)):*
> ```Go
> 	lookupCols columns
> 
> 	primaryFetcher     *sqlbase.RowFetcher
> ```

This block could use a comment

---

*[pkg/sql/distsqlrun/joinreader.go, line 125 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/25628#-LCmu3-6-lAemv8yVaCZ:-LCmu3-6-lAemv8yVaC_:b-j56edb) ([raw file](https://github.com/cockroachdb/cockroach/blob/27608f8ae412589cb9d6510cae3245fe62b9dad7/pkg/sql/distsqlrun/joinreader.go#L125)):*
> ```Go
> 	var neededCols util.FastIntSet
> 	if jr.index == &jr.desc.PrimaryIndex ||
> 		indexIsCovering(jr.index, jr.desc.ColumnIdxMap(), jr.out.neededColumns(), len(jr.inputTypes)) {
> ```

We are calculating this map multiple times (once here, and once in every call to initRowFetcher). We should just calculate it once and pass it initRowFetcher.

---

*[pkg/sql/distsqlrun/joinreader.go, line 127 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/25628#-LCmry6L819gG5-0y7g4:-LCmry6L819gG5-0y7g5:bjju6jm) ([raw file](https://github.com/cockroachdb/cockroach/blob/27608f8ae412589cb9d6510cae3245fe62b9dad7/pkg/sql/distsqlrun/joinreader.go#L127)):*
> ```Go
> 		indexIsCovering(jr.index, jr.desc.ColumnIdxMap(), jr.out.neededColumns(), len(jr.inputTypes)) {
> 		// jr.index includes all the needed output columns, so only need one lookup.
> 		neededCols = jr.rowFetcherColumns()
> ```

Add a comment for which it's clear how `rowFetcherColumns()` differs from `neededColumns()`. `rowFetchecColumns()` could also use a comment explaining the lookup join case

---

*[pkg/sql/distsqlrun/joinreader.go, line 181 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/25628#-LCmuh2V2L35H4s3OPMP:-LCmuh2V2L35H4s3OPMQ:biuwy2c) ([raw file](https://github.com/cockroachdb/cockroach/blob/27608f8ae412589cb9d6510cae3245fe62b9dad7/pkg/sql/distsqlrun/joinreader.go#L181)):*
> ```Go
> 		coveredCols.Add(numInputCols + colIdxMap[columnID])
> 	}
> 	return neededColumns.SubsetOf(coveredCols)
> ```

What about StoreColumnIDs?

---

*[pkg/sql/distsqlrun/joinreader.go, line 184 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/25628#-LCmuGuH90rOnufrn1MM:-LCmuGuH90rOnufrn1MN:btnj4fv) ([raw file](https://github.com/cockroachdb/cockroach/blob/27608f8ae412589cb9d6510cae3245fe62b9dad7/pkg/sql/distsqlrun/joinreader.go#L184)):*
> ```Go
> }
> 
> func getColumnIndices(
> ```

maybe getIndexColSet

---

*[pkg/sql/distsqlrun/joinreader.go, line 188 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/25628#-LCmuYKZ1-fvbrUvQ9Cx:-LCmuYKZ1-fvbrUvQ9Cy:bqn8xu8) ([raw file](https://github.com/cockroachdb/cockroach/blob/27608f8ae412589cb9d6510cae3245fe62b9dad7/pkg/sql/distsqlrun/joinreader.go#L188)):*
> ```Go
> ) util.FastIntSet {
> 	cols := util.MakeFastIntSet()
> 	for _, columnID := range index.ColumnIDs {
> ```

What about ExtraColumnIDs, StoredColumnIDs?

---

*[pkg/sql/logictest/testdata/logic_test/lookup_join, line 275 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/25628#-LCmvl2s-lvcncPko8Q7:-LCmvl2s-lvcncPko8Q8:bho6v0x) ([raw file](https://github.com/cockroachdb/cockroach/blob/27608f8ae412589cb9d6510cae3245fe62b9dad7/pkg/sql/logictest/testdata/logic_test/lookup_join#L275)):*
> ```
> 
> 
> ####################################
> ```

These tests don't verify that what the processor is doing internally is right. For example, it might be doing secondary lookups unnecessarily.

We should add a group of tests that use the KV TRACING facility to make sure we are doing the correct lookups (like https://github.com/cockroachdb/cockroach/blob/master/pkg/sql/logictest/testdata/logic_test/insert#L46). The tests should contain cases that break if we don't consider ExtraColumnIDs and StoreColumnIDs.

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/25628#-:-LCmuqS06_5YXYuXbnjT:b6zp48o)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: 1 of 9 files reviewed at latest revision, 9 unresolved discussions.

---

*[pkg/sql/distsql_physical_planner.go, line 802 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/25628#-LCjyGPF6xh1ohlaOGO9:-LCnOPON5pbZy_vXQ8St:bhz3rjp) ([raw file](https://github.com/cockroachdb/cockroach/blob/79776afe24b0ed005a2a04b5bebb4f3b1e813e23/pkg/sql/distsql_physical_planner.go#L802)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

We are putting whatever version of `n.desc` we have in the spec (above); `distsqlrun` does not look up table descriptors.
</blockquote></details>

Ah, got it. Agreed that this is safe. Carry on.

---

*[pkg/sql/distsql_physical_planner.go, line 1831 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/25628#-LCjyTwY1lK4cZ41eV6m:-LCnOW9j8RCUkGXXOPem:b48ow8x) ([raw file](https://github.com/cockroachdb/cockroach/blob/79776afe24b0ed005a2a04b5bebb4f3b1e813e23/pkg/sql/distsql_physical_planner.go#L1831)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

I don't have all the context here, but perhaps the idea was that the query can be rewritten to flip the order if necessary. If we want to handle the symmetric case, it should be a separate PR.
</blockquote></details>

Do we do any join associativity in the 2.0 planner? If not, we might want to document this restriction when we add documentation for `experimental_force_lookup_join`. 

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/25628)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Review status: 1 of 9 files reviewed at latest revision, 9 unresolved discussions.

---

*[pkg/sql/distsql_physical_planner.go, line 1831 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/25628#-LCjyTwY1lK4cZ41eV6m:-LCnQcPBCkHk97pmKpdh:bhryxu9) ([raw file](https://github.com/cockroachdb/cockroach/blob/79776afe24b0ed005a2a04b5bebb4f3b1e813e23/pkg/sql/distsql_physical_planner.go#L1831)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Do we do any join associativity in the 2.0 planner? If not, we might want to document this restriction when we add documentation for `experimental_force_lookup_join`. 
</blockquote></details>

We don't. Agreed on the documentation.

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/25628)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: 1 of 9 files reviewed at latest revision, 9 unresolved discussions.

---

*[pkg/sql/distsql_physical_planner.go, line 1831 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/25628#-LCjyTwY1lK4cZ41eV6m:-LCnRziH8a7UpKcMFjgJ:bddeiv0) ([raw file](https://github.com/cockroachdb/cockroach/blob/79776afe24b0ed005a2a04b5bebb4f3b1e813e23/pkg/sql/distsql_physical_planner.go#L1831)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

We don't. Agreed on the documentation.
</blockquote></details>

Cc @lhirata who I believe is documenting `experimental_force_lookup_join`. 

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/25628)*
<!-- Sent from Reviewable.io -->

asubiotto commented:
Thanks for taking this on, left some minor comments

---

Review status: 1 of 9 files reviewed at latest revision, 9 unresolved discussions.

---

*[pkg/sql/distsqlrun/joinreader.go, line 181 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/25628#-LCmuh2V2L35H4s3OPMP:-LCoB7zn7OZEDkDShcZ_:bkvaa77) ([raw file](https://github.com/cockroachdb/cockroach/blob/27608f8ae412589cb9d6510cae3245fe62b9dad7/pkg/sql/distsqlrun/joinreader.go#L181)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

What about StoreColumnIDs?
</blockquote></details>

This looks sort of similar to what is in the `scrub_tablereader.go`, you might be able to share the logic.

---

*[pkg/sql/distsqlrun/joinreader.go, line 443 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/25628#-LCoDmR1AlmK6v4mZh5l:-LCoDmR1AlmK6v4mZh5m:b-y2s4bi) ([raw file](https://github.com/cockroachdb/cockroach/blob/27608f8ae412589cb9d6510cae3245fe62b9dad7/pkg/sql/distsqlrun/joinreader.go#L443)):*
> ```Go
> ) (sqlbase.EncDatumRow, error) {
> 	index := jr.desc.PrimaryIndex
> 	keyValues := make(sqlbase.EncDatumRow, len(index.ColumnIDs))
> ```

This might benefit from using a scratch slice instead.

---

*[pkg/sql/distsqlrun/joinreader.go, line 453 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/25628#-LCoFmLhAeGYumu0aF1I:-LCoFmLhAeGYumu0aF1J:b608o9l) ([raw file](https://github.com/cockroachdb/cockroach/blob/27608f8ae412589cb9d6510cae3245fe62b9dad7/pkg/sql/distsqlrun/joinreader.go#L453)):*
> ```Go
> 	}
> 	spans := []roachpb.Span{{Key: key, EndKey: key.PrefixEnd()}}
> 	err = jr.primaryFetcher.StartScan(ctx, txn, spans, false, 0, false)
> ```

Add comments to arguments

---

*[pkg/sql/distsqlrun/joinreader.go, line 466 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/25628#-LCoFAgr2skLAFrVX6zS:-LCoFAgr2skLAFrVX6zT:b-3mbaby) ([raw file](https://github.com/cockroachdb/cockroach/blob/27608f8ae412589cb9d6510cae3245fe62b9dad7/pkg/sql/distsqlrun/joinreader.go#L466)):*
> ```Go
> 		return nil, errors.New("expected exactly one row but found none")
> 	}
> 	nextRow, _, _, err := jr.primaryFetcher.NextRow(ctx)
> ```

I'm not sure how necessary this assertion is. If you want to keep it, you might want to consider exposing `rowFetcher.kvEnd`.

---

*[pkg/sql/logictest/testdata/logic_test/lookup_join, line 275 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/25628#-LCmvl2s-lvcncPko8Q7:-LCoCp4e2Y_nm4qsmKP3:booaj0u) ([raw file](https://github.com/cockroachdb/cockroach/blob/27608f8ae412589cb9d6510cae3245fe62b9dad7/pkg/sql/logictest/testdata/logic_test/lookup_join#L275)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

These tests don't verify that what the processor is doing internally is right. For example, it might be doing secondary lookups unnecessarily.

We should add a group of tests that use the KV TRACING facility to make sure we are doing the correct lookups (like https://github.com/cockroachdb/cockroach/blob/master/pkg/sql/logictest/testdata/logic_test/insert#L46). The tests should contain cases that break if we don't consider ExtraColumnIDs and StoreColumnIDs.
</blockquote></details>

I don't think we plumb kv tracing down to distsql, this might need to be done in a separate PR.

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/25628#-:-LCoK40f9AP5EHRbP-pf:b3e47ec)*
<!-- Sent from Reviewable.io -->

solongordon commented:
Thanks all for the comments. I think I've addressed everything except for adding more granular unit testing. I also still have a TODO for batching the primary index fetches, which I'd like to do in a follow-up PR if no one objects.

---

Review status: 0 of 12 files reviewed at latest revision, 12 unresolved discussions.

---

*[pkg/sql/distsqlrun/joinreader.go, line 64 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/25628#-LCmtUX27OL1CxSDBIqt:-LD28R0y3-FXGMD1C3rF:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/27608f8ae412589cb9d6510cae3245fe62b9dad7/pkg/sql/distsqlrun/joinreader.go#L64)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

This block could use a comment
</blockquote></details>

Done.

---

*[pkg/sql/distsqlrun/joinreader.go, line 125 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/25628#-LCmu3-6-lAemv8yVaCZ:-LD2Gm9-2n0WNwtbm8Fx:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/27608f8ae412589cb9d6510cae3245fe62b9dad7/pkg/sql/distsqlrun/joinreader.go#L125)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

We are calculating this map multiple times (once here, and once in every call to initRowFetcher). We should just calculate it once and pass it initRowFetcher.
</blockquote></details>

Done.

---

*[pkg/sql/distsqlrun/joinreader.go, line 127 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/25628#-LCmry6L819gG5-0y7g4:-LD2LQD9CXAFE9cDFdmE:b5n9qlp) ([raw file](https://github.com/cockroachdb/cockroach/blob/27608f8ae412589cb9d6510cae3245fe62b9dad7/pkg/sql/distsqlrun/joinreader.go#L127)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Add a comment for which it's clear how `rowFetcherColumns()` differs from `neededColumns()`. `rowFetchecColumns()` could also use a comment explaining the lookup join case
</blockquote></details>

Done, hopefully clearer now. I renamed `rowFetcherColumns` to `neededRightCols` which is more reflective of its purpose.

---

*[pkg/sql/distsqlrun/joinreader.go, line 181 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/25628#-LCmuh2V2L35H4s3OPMP:-LD2FOAl4Zp65AsZT5T4:b-48vilg) ([raw file](https://github.com/cockroachdb/cockroach/blob/27608f8ae412589cb9d6510cae3245fe62b9dad7/pkg/sql/distsqlrun/joinreader.go#L181)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

This looks sort of similar to what is in the `scrub_tablereader.go`, you might be able to share the logic.
</blockquote></details>

Good call, done.

---

*[pkg/sql/distsqlrun/joinreader.go, line 184 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/25628#-LCmuGuH90rOnufrn1MM:-LD28gUdAX5D0SdycJjn:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/27608f8ae412589cb9d6510cae3245fe62b9dad7/pkg/sql/distsqlrun/joinreader.go#L184)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

maybe getIndexColSet
</blockquote></details>

Done.

---

*[pkg/sql/distsqlrun/joinreader.go, line 188 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/25628#-LCmuYKZ1-fvbrUvQ9Cx:-LD28hfI2wDTDPz0gque:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/27608f8ae412589cb9d6510cae3245fe62b9dad7/pkg/sql/distsqlrun/joinreader.go#L188)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

What about ExtraColumnIDs, StoredColumnIDs?
</blockquote></details>

Done.

---

*[pkg/sql/distsqlrun/joinreader.go, line 443 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/25628#-LCoDmR1AlmK6v4mZh5l:-LD29-vN4K4PT8es1Pwj:b-15efoz) ([raw file](https://github.com/cockroachdb/cockroach/blob/27608f8ae412589cb9d6510cae3245fe62b9dad7/pkg/sql/distsqlrun/joinreader.go#L443)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

This might benefit from using a scratch slice instead.
</blockquote></details>

Could you clarify? Unless I'm misunderstanding I think this is already a scratch slice. (`EncDatumRow` is just `[]EncDatum`).

---

*[pkg/sql/distsqlrun/joinreader.go, line 453 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/25628#-LCoFmLhAeGYumu0aF1I:-LD29ORN1P4WUawx2c01:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/27608f8ae412589cb9d6510cae3245fe62b9dad7/pkg/sql/distsqlrun/joinreader.go#L453)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

Add comments to arguments
</blockquote></details>

Done.

---

*[pkg/sql/distsqlrun/joinreader.go, line 466 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/25628#-LCoFAgr2skLAFrVX6zS:-LD2IgFD7_G8yf0girhD:b-48pgob) ([raw file](https://github.com/cockroachdb/cockroach/blob/27608f8ae412589cb9d6510cae3245fe62b9dad7/pkg/sql/distsqlrun/joinreader.go#L466)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

I'm not sure how necessary this assertion is. If you want to keep it, you might want to consider exposing `rowFetcher.kvEnd`.
</blockquote></details>

Yeah it's not strictly necessary, but it's cheap and if it fails we're probably returning incorrect results, so I thought it was worth it.

---

*[pkg/sql/logictest/testdata/logic_test/lookup_join, line 275 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/25628#-LCmvl2s-lvcncPko8Q7:-LD2OPyTE9MKMppcqbAV:b-1dgfi8) ([raw file](https://github.com/cockroachdb/cockroach/blob/27608f8ae412589cb9d6510cae3245fe62b9dad7/pkg/sql/logictest/testdata/logic_test/lookup_join#L275)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

I don't think we plumb kv tracing down to distsql, this might need to be done in a separate PR.
</blockquote></details>

Maybe this can be addressed in `joinreader_test.go`? Going to take a look at supplementing that now.

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/25628#-:-LD2POjIEp0x4b1MMk7C:b-eky68b)*
<!-- Sent from Reviewable.io -->

asubiotto commented:
Review status: 0 of 12 files reviewed at latest revision, 12 unresolved discussions, some commit checks failed.

---

*[pkg/sql/distsqlrun/joinreader.go, line 443 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/25628#-LCoDmR1AlmK6v4mZh5l:-LD2lwNYELliPlN3g-cA:bezet24) ([raw file](https://github.com/cockroachdb/cockroach/blob/27608f8ae412589cb9d6510cae3245fe62b9dad7/pkg/sql/distsqlrun/joinreader.go#L443)):*
<details><summary><i>Previously, solongordon (Solon) wrote…</i></summary><blockquote>

Could you clarify? Unless I'm misunderstanding I think this is already a scratch slice. (`EncDatumRow` is just `[]EncDatum`).
</blockquote></details>

On every lookup, you're creating a new `keyValues` slice. I'm suggesting you have one on the `joinReader` called `scratchKeyValues` that you `make` once then reset on each lookup similar to https://github.com/cockroachdb/cockroach/blob/cc4511e1be866da968b94432d7537cf2bba5af3b/pkg/sql/distsqlrun/aggregator.go#L726.

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/25628)*
<!-- Sent from Reviewable.io -->

solongordon commented:
Review status: 0 of 12 files reviewed at latest revision, 12 unresolved discussions, some commit checks failed.

---

*[pkg/sql/distsqlrun/joinreader.go, line 443 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/25628#-LCoDmR1AlmK6v4mZh5l:-LD3BBLQ9SPfTpiPommd:b-mjk6mi) ([raw file](https://github.com/cockroachdb/cockroach/blob/27608f8ae412589cb9d6510cae3245fe62b9dad7/pkg/sql/distsqlrun/joinreader.go#L443)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

On every lookup, you're creating a new `keyValues` slice. I'm suggesting you have one on the `joinReader` called `scratchKeyValues` that you `make` once then reset on each lookup similar to https://github.com/cockroachdb/cockroach/blob/cc4511e1be866da968b94432d7537cf2bba5af3b/pkg/sql/distsqlrun/aggregator.go#L726.
</blockquote></details>

Ah got it, thanks. If it's cool with you I think I'll hold off on the scratch slice for now since this should be a small array which is allocated on the stack. Also this code is likely to change when I add batching.

---

*[pkg/sql/logictest/testdata/logic_test/lookup_join, line 275 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/25628#-LCmvl2s-lvcncPko8Q7:-LD3BcUoCj4raQdfOK4R:b-4zfi3g) ([raw file](https://github.com/cockroachdb/cockroach/blob/27608f8ae412589cb9d6510cae3245fe62b9dad7/pkg/sql/logictest/testdata/logic_test/lookup_join#L275)):*
<details><summary><i>Previously, solongordon (Solon) wrote…</i></summary><blockquote>

Maybe this can be addressed in `joinreader_test.go`? Going to take a look at supplementing that now.
</blockquote></details>

I added a unit test which covers the various cases here (ColumnID, ExtraColumnID, StoreColumnID, and non-covering index.)

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/25628)*
<!-- Sent from Reviewable.io -->

solongordon commented:
Review status: 0 of 16 files reviewed at latest revision, 7 unresolved discussions.

---

*[pkg/sql/distsqlrun/joinreader.go, line 443 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/25628#-LCoDmR1AlmK6v4mZh5l:-LD7tN6XAiidDZf87_2W:b-j0t85i) ([raw file](https://github.com/cockroachdb/cockroach/blob/27608f8ae412589cb9d6510cae3245fe62b9dad7/pkg/sql/distsqlrun/joinreader.go#L443)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Ok to hold off, but wanted to mention that `make` never allocates on the stack AFAIK
</blockquote></details>

Hm, I came across this thread which seems to indicate otherwise, but I'd love to know more. https://groups.google.com/d/msg/golang-nuts/KdbtOqNK6JQ/ehUOmI7LaKwJ

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/25628)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: 0 of 16 files reviewed at latest revision, 7 unresolved discussions.

---

*[pkg/sql/distsqlrun/joinreader.go, line 443 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/25628#-LCoDmR1AlmK6v4mZh5l:-LD7vBrz8RzR3_ljX1fs:b-wnuepw) ([raw file](https://github.com/cockroachdb/cockroach/blob/27608f8ae412589cb9d6510cae3245fe62b9dad7/pkg/sql/distsqlrun/joinreader.go#L443)):*
<details><summary><i>Previously, solongordon (Solon) wrote…</i></summary><blockquote>

Hm, I came across this thread which seems to indicate otherwise, but I'd love to know more. https://groups.google.com/d/msg/golang-nuts/KdbtOqNK6JQ/ehUOmI7LaKwJ
</blockquote></details>

My understanding is that Dave Cheney is wrong there and that `make` always results in a heap allocation. I'd have to test that to be certain. I certainly wouldn't rely on the compiler recognizing that a slice is not escaping and can be allocated on the stack.

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/25628)*
<!-- Sent from Reviewable.io -->

solongordon commented:
bors r+

solongordon commented:
Review status: 0 of 16 files reviewed at latest revision, 7 unresolved discussions, all commit checks successful.

---

*[pkg/sql/distsqlrun/joinreader.go, line 443 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/25628#-LCoDmR1AlmK6v4mZh5l:-LD86Erb3QRB9iQqcSX7:bs35opb) ([raw file](https://github.com/cockroachdb/cockroach/blob/27608f8ae412589cb9d6510cae3245fe62b9dad7/pkg/sql/distsqlrun/joinreader.go#L443)):*
<details><summary><i>Previously, solongordon (Solon) wrote…</i></summary><blockquote>

I did an experiment which appears to back up Dave. As far as I can tell the slice is allocated on the heap when it exceeds 8192 bytes, which I suppose is the stack frame size.
```
solon:~/tmp/stackallocation$ cat main_test.go
package main

import "testing"

func BenchmarkSlice8191(b *testing.B) {
	for i := 0; i < b.N; i++ {
		s := make([]int, 8191)
		s[0] = 1
	}
}

func BenchmarkSlice8192(b *testing.B) {
	for i := 0; i < b.N; i++ {
		s := make([]int, 8192)
		s[0] = 1
	}
}

solon:~/tmp/stackallocation$ go tool compile -m main_test.go
main_test.go:5:28: BenchmarkSlice8191 b does not escape
main_test.go:7:12: BenchmarkSlice8191 make([]int, 8191) does not escape
main_test.go:14:12: make([]int, 8192) escapes to heap
main_test.go:12:28: BenchmarkSlice8192 b does not escape

solon:~/tmp/stackallocation$ go test -bench=.
goos: darwin
goarch: amd64
BenchmarkSlice8191-4   	 1000000	      1248 ns/op
BenchmarkSlice8192-4   	  300000	      4222 ns/op
PASS
ok  	_/Users/solon/tmp/stackallocation	2.583s
```
</blockquote></details>

The plot thickens though: the compiler isn't smart enough to allocate on the stack if the length is not a constant. I think that means it will allocate on the heap in this case after all.

https://github.com/golang/go/issues/20533

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/25628)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: 0 of 16 files reviewed at latest revision, 7 unresolved discussions, all commit checks successful.

---

*[pkg/sql/distsqlrun/joinreader.go, line 443 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/25628#-LCoDmR1AlmK6v4mZh5l:-LD87lClAyTIO4jx-xNO:bmir1wa) ([raw file](https://github.com/cockroachdb/cockroach/blob/27608f8ae412589cb9d6510cae3245fe62b9dad7/pkg/sql/distsqlrun/joinreader.go#L443)):*
<details><summary><i>Previously, solongordon (Solon) wrote…</i></summary><blockquote>

The plot thickens though: the compiler isn't smart enough to allocate on the stack if the length is not a constant. I think that means it will allocate on the heap in this case after all.

https://github.com/golang/go/issues/20533
</blockquote></details>

Thanks for following up on this. TIL that the compiler will allocate constant length slices on the stack in some instances. That's interesting. 

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/25628)*
<!-- Sent from Reviewable.io -->
